service/matching/matching_engine.go

```diff
@@ -837,6 +837,155 @@ func (e *matchingEngineImpl) listTaskQueuePartitions(request *matchingservice.Li
 	return partitionHostInfo, nil
 }
 
+func (e *matchingEngineImpl) UpdateWorkerVersioningRules(
+	ctx context.Context,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest,
+) (*workflowservice.UpdateWorkerVersioningRulesResponse, error) {
+	ns, err := e.namespaceRegistry.GetNamespace(namespace.Name(req.GetNamespace()))
+	if err != nil {
+		return nil, err
+	}
+
+	taskQueue, err := newTaskQueueID(ns.ID(), req.GetTaskQueue(), enumspb.TASK_QUEUE_TYPE_WORKFLOW)
+	if err != nil {
+		return nil, err
+	}
+
+	tqMgr, err := e.getTaskQueuePartitionManager(ctx, taskQueue, normalStickyInfo, true)
+	if err != nil {
+		return nil, err
+	}
+
+	updateOptions := UserDataUpdateOptions{}
+	operationCreatedTombstones := false
+
+	// todo: possibly do some limit checking here? does that prevent a race maybe?
```

Comment:
thinking just checking the limits in InsertAssignmentRule and InsertCompatibleRedirectRule methods should be enough

Reply:
for sure. I just wrote that because for some reason, in the existing `UpdateWorkerBuildIdCompatibility` function, they do some operations to update the `UserDataUpdateOptions` before going into the `UpdateUserData` function, and I'm not sure why, or whether I need to do something similar.

It's in lines 1009 - 1015:
```
	switch req.GetOperation().(type) {
	case *matchingservice.UpdateWorkerBuildIdCompatibilityRequest_ApplyPublicRequest_:
		// Only apply the limit when request is initiated by a user.
		updateOptions.TaskQueueLimitPerBuildId = e.config.TaskQueueLimitPerBuildId(ns.Name().String())
	case *matchingservice.UpdateWorkerBuildIdCompatibilityRequest_RemoveBuildIds_:
		updateOptions.KnownVersion = req.GetRemoveBuildIds().GetKnownUserDataVersion()
	}
```

Reply:
maybe @dnr can clarify